File: 07-03-viewmodel/java-07-03-viewmodel/app/src/main/java/com/example/viewmodel/PokeViewModel.java
package com.example.viewmodel;

import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MutableLiveData;
import android.arch.lifecycle.ViewModel;

import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class PokeViewModel extends ViewModel {
    private MutableLiveData<List<Result>> results;
    private PokeAPI pokeAPI;

    public LiveData<List<Result>> getResults() {
        if (results == null) {
            results = new MutableLiveData<>();

            Retrofit retrofit = new Retrofit.Builder()
                    .baseUrl("http://pokeapi.co/api/v2/")
                    .addConverterFactory(GsonConverterFactory.create())
                    .build();

            pokeAPI = retrofit.create(PokeAPI.class);
            loadResults();
        }
        return results;
    }

    private void loadResults() {
        pokeAPI.listPokemons().enqueue(new Callback<Response>() {
            @Override
            public void onResponse(Call<Response> call, retrofit2.Response<Response> response) {
                results.setValue(response.body().results);
            }

            @Override
            public void onFailure(Call<Response> call, Throwable t) {
            }
        });
    }
}

File: 07-03-viewmodel/kotlin-07-03-viewmodel/app/src/main/java/com/example/viewmodel/PokeViewModel.kt
package com.example.viewmodel

import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory

class PokeViewModel : ViewModel() {
    val results: MutableLiveData<List<Result>> by lazy {
        var liveData = MutableLiveData<List<Result>>()
        loadResults()
        liveData
    }
    private val pokeAPI: PokeAPI by lazy {
        val retrofit = Retrofit.Builder()
                .baseUrl("https://pokeapi.co/api/v2/")
                .addConverterFactory(GsonConverterFactory.create())
                .build()

        retrofit.create(PokeAPI::class.java)
    }

    private fun loadResults() {
        pokeAPI.listPokemons().enqueue(object : Callback<Response> {
            override fun onResponse(call: Call<Response>, response: retrofit2.Response<Response>) {
                results.value = response.body()?.results
            }

            override fun onFailure(call: Call<Response>, t: Throwable) {}
        })
    }
}
